packages/react-reconciler/src/ReactFiberHotReloading.js

```diff
@@ -0,0 +1,258 @@
+    return family.currentType;
+  } else {
+    return type;
+  }
+}
+
+export function isCompatibleFamilyForHotReloading(
+  fiber: Fiber,
+  element: ReactElement,
+): boolean {
+  if (__DEV__) {
+    if (familiesByType === null) {
+      // Hot reloading is disabled.
+      return false;
+    }
+
+    const prevType = fiber.elementType;
+    const nextType = element.type;
+    // If we got here, we know types aren't === equal.
+    let needsCompareFamilies = false;
+    switch (fiber.tag) {
+      case FunctionComponent: {
+        if (typeof nextType === 'function') {
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      case ForwardRef: {
+        if (
+          typeof nextType === 'object' &&
+          nextType !== null &&
+          nextType.$$typeof === REACT_FORWARD_REF_TYPE
+        ) {
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      case MemoComponent:
+      case SimpleMemoComponent: {
+        if (
+          typeof nextType === 'object' &&
+          nextType !== null &&
+          nextType.$$typeof === REACT_MEMO_TYPE
+        ) {
+          // TODO: if it was but can no longer be simple,
+          // we shouldn't set this.
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      // TODO: maybe support lazy?
+      default:
+        return false;
+    }
+
+    // Check if both types have a family and it's the same one.
+    if (needsCompareFamilies) {
+      // Note: memo() and forwardRef() we'll compare outer rather than inner type.
+      // This means both of them need to be registered to preserve state.
+      // If we unwrapped and compared the inner types for wrappers instead,
+      // then we would risk falsely saying two separate memo(Foo)
+      // calls are equivalent because they wrap the same Foo function.
+      const prevFamily = familiesByType.get(prevType);
+      if (
+        prevFamily !== undefined &&
+        prevFamily === familiesByType.get(nextType)
+      ) {
+        return true;
+      }
+    }
+    return false;
+  } else {
+    return false;
+  }
+}
+
+export function markFailedErrorBoundaryForHotReloading(fiber: Fiber) {
+  if (__DEV__) {
+    if (familiesByType === null) {
+      // Not hot reloading.
+      return;
+    }
+    if (typeof WeakSet !== 'function') {
+      return;
+    }
+    if (failedBoundaries === null) {
+      failedBoundaries = new WeakSet();
+    }
+    failedBoundaries.add(fiber);
+  }
+}
+
+export function scheduleHotUpdate(hotUpdate: HotUpdate): void {
+  if (__DEV__) {
+    // TODO: warn if its identity changes over time?
+    familiesByType = hotUpdate.familiesByType;
+
+    const {root, staleFamilies, updatedFamilies} = hotUpdate;
+    flushPassiveEffects();
+    flushSync(() => {
+      scheduleFibersWithFamiliesRecursively(
+        root.current,
+        updatedFamilies,
+        staleFamilies,
+      );
+    });
+    flushPassiveEffects();
```

Comment:
What's this for?

Reply:
I can remove. I figured I want to do as much as I can related to the update synchronously but I don't have any particular justification except that the developer is waiting for reload and its effects to finish. I guess it's fair we don't want to deviate from real behavior too much.

Reply:
Removed